### ISSUE TITLE
New Task: Release v1.0 of ATN DAC netCDF profile template

### ISSUE THREAD

**Issue by MathewBiddle** [ioos/marine_life_data_network#57]
### Who is requesting this?

@ioos/marine-life and NCEI

### What is being requested?

ATN establishes a netCDF template for profile data. This includes:

- [ ] Draft the netCDF template following the contributing guidelines published from https://github.com/ioos/ioos-atn-data/issues/79
- [ ] Present netCDF template to NCEI for approval (maybe needs NCEI Dataset Readiness Review (DSRR) board approval?)

### What is the requested deadline and why?

2024-12-31

### What is the current status quo (i.e., what happens if this does not get done)?

NCEI will not archive data that aren't fully documented. The telemetry community will splinter into multiple file templates for similar observing methods.

### What indicates this is done (i.e., how do we know this is complete)?

- [ ]  A successfully merged pull request to the jekyll base webpage at https://ioos.github.io/ioos-atn-data/ to document v1.0 iteration of the profile.

### Provide a description or any other important information.

xref:
* https://github.com/ioos/marine_life_data_network/issues/42

Split from:
* https://github.com/ioos/marine_life_data_network/issues/16

**Comment by MathewBiddle** [comment]
This should build heavily on the work done for satellite trajectory work. So, we don't expect this to be a significant lift, but may take time to get to.

**Comment by mmckinzie** [comment]
Agreed. We're still working on finalizing the related trajectory template and pushing relevant updates to the ATN satellite telemetry [specification v.1 website] (https://ioos.github.io/ioos-atn-data/atn-sat-telem-specification-v1-0.html). There are also some lingering issues to resolve regarding select global attributes (e.g., :Acknowledgements, :sea names) within the files themselves and related to corresponding metadata harvesting/generation by NCEI. Once everything within the trajectory template is finalized and aligns with NCEI documentation, we'll move on to the profile template specification documentation, website integration and NCEI approval. Axiom also still needs to work on automating and operationalizing the NCEI pipeline.

**Comment by MathewBiddle** [comment]
@mmckinzie I think you, @laurabrenskelle, and I can come up with what the template should look like and hand that over to Axiom to implement. The goal of this issue is to establish what the template will look like. 

I am going to set a relatively aggressive timeline of having the first version of the template released by the end of the calendar year (Jan 31st). I think it's doable as we've already done a lot of the heavy lifting on the trajectory template.

**Comment by mmckinzie** [comment]
@MathewBiddle its more or less already done. No need to come up with one. We just haven't passed it off to Axiom because we're still trying to resolve issues with the trajectory template

**Comment by MathewBiddle** [comment]
We need to be clear on what means done in this case. 

My definition of done is a `.cdl` file that is added to https://github.com/ioos/ioos-atn-data/tree/gh-pages/_docs/data/templates which clearly defines what the netCDF file will look like for the profile data. That should be the "thing" that is passed off to Axiom for implementation.

Additionally, the documentation page (https://ioos.github.io/ioos-atn-data/) should be updated to clearly describe the specification. Similar to the trajectory page https://ioos.github.io/ioos-atn-data/atn-sat-telem-specification-v1-0.html.

We might need to rethink how the webpage is organized or named to be clear about which template is which.

**Comment by MathewBiddle** [comment]
See my update to the contributing guide for details. https://github.com/ioos/ioos-atn-data/blob/gh-pages/CONTRIBUTING.md#file-template-development

**Comment by mmckinzie** [comment]
Fair enough. We are not yet to the point where a .cdl file could be generated and added the ioos-atn-data github nor shared with Axiom and I agree that eventually the documentation page should be updated to describe specification and yes, I'd appreciate your help generating the cdl file.

I have been working on generating a 'template' that could be used as the basis the generate the .cdl file and provide context to specification on website which includes the complete list of global attributes including their scope/source convention, disposition, content source, example and descriptions, and a list of variables and their associated attributes, values, source and examples for both WC and SMRU. 

I only have the global attributes finalize, still working on adding the variables. That file is available [here](https://docs.google.com/spreadsheets/d/1lNHa-JewCE1zgOiLm6lNp8gstPjug8xb/edit?usp=sharing&ouid=113417505623695767850&rtpof=true&sd=true). Please do not share this file publicly as it contains embargoed/protected information.

**Comment by MathewBiddle** [comment]
Things are changing with respect to requirements to get data into NCEI. We have some options right now.

1. _Preferred_ Create netCDF files to submit to NCEI. 
   1. If we have the netCDF template and the work is already done to make those files let's roll with them. I can update the current trajectory template to reflect the additional profile data. It's just a matter of how complex the backend is at Axiom to create said files.
1. Submit source data files to NCEI
   1. If possible, can you share the source files as they stand at ATN? If they are well documented enough, we might be able to push those directly to NCEI. Instead of generating netCDF packages. We'd still need to do the checksum stuff, however.

**Comment by MathewBiddle** [comment]
Okay, the profile template is exactly the same as the trajectory template. See my comment at https://github.com/ioos/ioos-atn-data/pull/106

See the template at https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template.cdl

**Comment by MathewBiddle** [comment]
The [satellite profile template](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template.cdl) is ready for review @mmckinzie 

I would like the template to be completed by April 30th. 

Once @mmckinzie approves, @iwensu0313 can hand over to tech team to implement and start building example files.

**Comment by MathewBiddle** [comment]
@mmckinzie please review the template linked above. It should line up well with the trajectory template. We're just adding geophysical variables and actual values for depth.

**Comment by mmckinzie** [comment]
@MathewBiddle, I will try to dig into this later this week. But don't foresee any reason why it cannot be finalized by April 30th.

**Comment by mmckinzie** [comment]
@MathewBiddle, due to an unforeseen family emergency that required me to be OOO for several weeks, I have not yet had a chance to fully review the profile template. But I hope to get back to you with comments by early next week. However, transitioning from the trajectory template to the profile template will be slightly more involved in practice than simply adding the geophysical variables and actual depth, partly because it requires accommodating data files from additional tag manufacturers. More soon.

**Comment by mmckinzie** [comment]
@MathewBiddle So I finally had a chance to start digging into the [profile template cdl](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template.cdl) you drafted based on the trajectory template. I wasn't sure how best to provide feedback, and I didn't want to edit what you created directly, so I drafted 2 updated versions: [MKM_1](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template_MKM_1.cdl) and [MKM_2](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template_MKM_2.cdl). 

In the first round of edits ([MKM_1](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template_MKM_1.cdl)), I essentially left the cdl you created as is, but added or edited variable notes, updated select variable attributes, added a few missing global attributes, and revised the listed SourceConventions and/or descriptions.  

However, there are numerous variables specific to Wildlife Computer data files within your original profile template cdl and therefore, also within my first edited version. Many of these variables are also present only in the WC trajectory files, not in the corresponding WC profile files, and will, of course, not be included in the SMRU trajectory and/or profile files. So, in my second round of edits ([MKM_2](https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template_MKM_2.cdl)), I removed those variables to list only the variables we can expect to find in the profile files, regardless of manufacturer. 

Ultimately, I suspect that Tetra Tech will need a cdl specifically for WC, another for SMRU profile files (and one for SMRU trajectory files), each including the additional manufacturer-specific variables, their corresponding lists of variable attributes, and the generalized container variables more explicitly laid out. The required global attributes will remain consistent regardless of manufacturer, although certain elements will need to be specified to ensure what is reported in a specific global attribute is appropriate (e.g., : acknowledgment; we will either need to update standard text to remove Wildlife Computers or have a standard version for each manufacturer). I will work on these versions next.

**Comment by MathewBiddle** [comment]
Thanks @mmckinzie for taking a look! 

After tagging up with @laurabrenskelle it sounds like you all have thoughts about a higher level template that can accommodate multiple manufacturers. I think that is MKM_2? After comparing your two templates, the differences make sense. Your expertise is appreciated.

Keep in mind, in the template we can have **core** information that is the same across all manufacturers and we can have **ancillary** information that might be specific to one manufacturer/deployment. The ancillary information can be in the form of attribute metadata or data variables and can come and go as appropriate. The **core** information would be the information that is consistent regardless of manufacturer/deployment and would most likely be the information NCEI collects metadata from. 

RE: `:acknowledgment` let's come up with some standardized text that can be used across manufacturers.

**Comment by mmckinzie** [comment]
@MathewBiddle Yes, MKM_2 is the higher-level profile template that could accommodate multiple manufacturers and contains only the core variables. I can clean up the global attributes a bit to ensure they are more standardized.  I can also generate a trajectory template cdl that only contains core variables and standardized global attributes.

I'm also happy to work on generating versions that are more manufacturer-specific and include the additional ancillary information. I suspect these will be useful for Tetra Tech when building the actual pipeline. 

But the approach you outlined above regarding core vs. ancillary information seems logical to me.

**Comment by MathewBiddle** [comment]
@laurabrenskelle and I reviewed and it looks good. 

@akrelling @ksauby - can you provide a red flag review of https://github.com/ioos/ioos-atn-data/blob/gh-pages/_docs/data/templates/atn_satellite_profile_template_MKM_2.cdl ?

Only looking for major issues with the template before moving forward.

**Comment by akrelling** [comment]
@MathewBiddle , I took a look and these are my 2 cents:

- this template inherits from CF, ACDD, IOOS 1.2, NCEI template, and [ATN specs](https://ioos.github.io/ioos-atn-data/atn-sat-telem-specification-v1-0.html).
- from the IOOS 1.2 and NCEI profile specs, it seems like the dimension trajectory is required. Dimension z might also be required.  It's just not very clear whether these dimensions are required because there can be two types of files: for trajectory only and for profiles only. Maybe these requirements (z AND trajectory dimensions) are specific for files that combine profiles and trajectory into one.
- a few variables that are present in the ATN DAC specs don't seem to be present in this template, namely  “type”, "count", "comment", "instrument_location". Maybe they are not required.
- coverage_content_type attr for the crs variable is set as "referenceInformation" and should be “coordinate”
- the following attributes are missing for the instrument_container_var, according to ATN specs:  coordinates, comment, instrument, platform.  
- there's a typo in the "Conventions" global attribute. Right now it reads "Coventions."
- NOAA is strongly suggesting that an open data license is set as value for the "license" attribute. Example of values would be "https://creativecommons.org/publicdomain/zero/1.0/" for CC0, or "https://creativecommons.org/licenses/by/4.0/" for CC-BY. That would also promote usage of the data by more users.
- the NCO global attribute, required by ATN specs, is missing.

These are my considerations.
I do understand, however, that some (or maybe many) of these may not apply, due to the type of file. 
I hope this helps.

**Comment by MathewBiddle** [comment]
Thank you @akrelling ! This is great feedback. We will review and make adjustments as necessary.

**Comment by mmckinzie** [comment]
@akrelling Thanks for your review and comments.

- z dimension is in both the trajectory and profile file, at least within the variables. We can add the appropriate fields to the trajectory templates' global attributes if necessary, but for the trajectory files, these are surface measurements, so Z is always hardcoded to 0, whereas in the profile, an actual depth value is reported.
- Yes, not all of the variables present in the current trajectory template will be in the profile template and vice versa; this is either due to them being auxiliary variables instead of core variables, so removed from the manufacturer agnostic profile template or because they are simply not contained in the profile data file. 
- I'll update the profile template cdl to fix the other missing attributes, typos, etc. Thanks for pointing them out. As well as ensure those elements are correct within the yet to be generated manufacturer agonistic trajectory template cdl. 
- I will also remove the ATN license statement and update the attribute to list a specific open data license.  I'd lean towards CC-BY, but I'm fine with either.

I will ping you and @MathewBiddle once the updated manufacturer agonstic CDL file, trajectory and profile are ready for review.

**Comment by MathewBiddle** [comment]
Can @iwensu0313 help here in some way?
